docs/blockentities/container.md

```diff
@@ -0,0 +1,302 @@
+
+```java
+// Get the first item in the container.
+ItemStack stack = container.getItem(0);
+
+// Set the first item in the container to dirt.
+container.setItem(0, new ItemStack(Items.DIRT));
+
+// Removes a quantity of (up to) 16 from the third slot.
+container.removeItem(2, 16);
+```
+
+:::warning
+A container may throw an exception if trying to access a slot that is beyond its container size. Alternatively, they may return `ItemStack.EMPTY`, as is the case with (for example) `SimpleContainer`.
+:::
+
+## `Container`s on `ItemStack`s
+
+Until now, we mainly discussed `Container`s on `BlockEntity`s. However, they can also be applied to [`ItemStack`s][itemstack] using the `minecraft:container` [data component][datacomponent]:
+
+```java
+// We use SimpleContainer as the superclass here so we don't have to reimplement the item handling logic ourselves.
+// Due to implementation details of SimpleContainer, this may lead to race conditions if multiple parties
+// can access the container at the same time, so we're just going to assume our mod doesn't allow that.
+// You may of course use a different implementation of Container (or implement Container yourself) if needed.
+public class MyBackpackContainer extends SimpleContainer {
+    // The item stack this container is for. Passed into and set in the constructor.
+    private final ItemStack stack;
+    
+    public MyBackpackContainer(ItemStack stack) {
+        // We call super with our desired container size.
+        super(27);
+        // Setting the stack field.
+        this.stack = stack;
+        // We load the container contents from the data component (if present), which is represented
+        // by the ItemContainerContents class. If absent, we use ItemContainerContents.EMPTY.
+        ItemContainerContents contents = stack.getOrDefault(DataComponents.CONTAINER, ItemContainerContents.EMPTY);
+        // Copy the data component contents into our item stack list.
+        contents.copyInto(this.getItems());
+    }
+
+    // When the contents are changed, we save the data component on the stack.
+    @Override
+    public void setChanged() {
+        super.setChanged();
+        this.stack.set(DataComponents.CONTAINER, ItemContainerContents.fromItems(this.getItems()));
+    }
+}
+```
+
+And voilà, you have created an item-backed container! Simply call `new MyBackpackContainer(stack)` to create a container for a menu or other use case.
+
+## `Container`s on `Entity`s
+
+`Container`s on `Entity`s are finicky: whether an entity has a container or not cannot be universally determined. It all depends on what entity you are handling, and as such can require a lot of special-casing.
+
+If you are creating an entity yourself, there is nothing stopping you from implementing `Container` on it directly, though be aware that you will not be able to use superclasses such as `SimpleContainer` (since `Entity` is the superclass).
+
+### `Container`s on `Mob`s
```

Comment:
This should also mention the InventoryCarrier interface which is implemented by some of the vanilla entities (villagers, pillagers, allays, piglins and wandering traders).